### PR TITLE
Added basic grpc interface for reinforcement learning

### PR DESCRIPTION
**Pull request by brollb**
This PR adds a basic RL interface (over rpc) which enables configuring scenarios, sending game engine commands to 0 AD, and executing the engine in lock step.

### CODE REVIEW

**Comment by na-Itms** [comment]
Hello, and thanks a lot for this very interesting pull request :+1: 

Unfortunately we do not take pull requests on GitHub, we use Phabricator for reviews. Please read [this wiki page](https://trac.wildfiregames.com/wiki/SubmittingPatches) to see how to submit a patch to [our Phabricator instance](https://code.wildfiregames.com/).

You can keep using git for local development without issue, but we will include patches upstream on Subversion, so you should think of a way to rebase and squash your work into sensible and atomic commits.

I would suggest:
- putting your bugfixes together and try to get them included first to get a better idea of our review process
- putting your needed refactoring together
- preparing a commit for the feature and keep the possible extra improvements for later

At a first glance, I also have a few comments on the code organization:
- the Python and generation script should go to a new folder in source/tools/
- the generated code should be generated as part of the build process, in premake

And finally, I would like to know whether you think this work could be easily built upon to create a new AI for the game of to improve our current one :slightly_smiling_face: 

Thanks again for your contribution. I'll close the PR when I see some news on the Phabricator side, don't hesitate to ask for help here as well.

**Comment by dmenager** [comment]
I am very interested in this pull request. In 2015, I worked on a project to extract the game state and send it to an artificial agent over TCP. The agent wasn't able to play the game, but through its observations, was able to make predictions about which player would win the game. I haven't looked into how you've decided to represent the state in your code, but I am very excited to know you're doing something like this! Please keep it going!

**Comment by brollb** [comment]
Thanks for the feedback!

Ironically, I meant to make this PR to my own fork and was still planning on doing a little more cleanup before contributing it back. The feedback and suggestions are much appreciated and I will make those changes! :smile: 

As far as the potential for building upon this work to create a new AI for the game, I had been imagining this as a way to make it easy to explore machine learning techniques and train models for research purposes and/or incorporation in the game. Although I don't imagine grpc being used for the final integration of a new AI in the game, I think it should make it much easier for experimentation into new or improved AI as the AI can be implemented using many different languages/libraries/etc.

@dmenager - That sounds really cool. Do you have more information about it? I am interested in learning more about your approach, accuracy, etc! As far as the game state goes, I am using the [output of GetFullRepresentation](https://github.com/brollb/0ad/blob/rl-interface/source/rlinterface/RLInterface.cpp#L182) used by the existing AI (though I am not [adding the passability/territory](https://github.com/0ad/0ad/blob/master/source/simulation2/components/CCmpAIManager.cpp#L1085-L1115) info currently).

**Comment by dmenager** [comment]
@brollb 
Yea, so you can checkout my fork of the repo. I was working on this project as a capstone project at the end of my undergrad four years ago. My team made their own `sendState` function. You can check it out here. https://github.com/dmenager/0ad/blob/master/source/ps/Game.cpp#L379 I think had we known about `GetFullRepresentation`, that would have saved us a lot of time.

Also, checkout my repo [INTENT-R](https://github.com/dmenager/INTENT-R). That was where we were running machine learning algorithms to try to recognize who would win the game. We originally had lofty goals of being able to recognize and predict other players' plans/intents, hence the name INTENT-R lol. But as the semester went on, we had to scale back quite a bit. So, go ahead and check it out!

At the end of the day, we were able observe AI-only games and predict who would win after 20 minutes of play with over 90% accuracy. We used hierarchical clustering to assign classes to the states, and then we used a support vector machine to classify observed states into the learned classes. 

It was a fun project, but we never got around to picking it back up. When you look at the code, you'll see that I was implementing an RL agent too, but I never got around to finishing it. And I don't think the code was very efficient/correct by the time the semester ended. So, I'm not sure how helpful it will be to you, but hopefully you enjoy!

**Comment by brollb** [comment]
@mhauskn - thanks for the feedback! When the GRPC code is generated from the protobuf file, `ServerContext*` is included in the [(virtual) method signatures](https://github.com/0ad/0ad/pull/25/files#diff-7ab8a5dfc3c2358c6cb76606ba0ce47cR153) (despite not being needed in the actual methods).

Yeah, this contribution does not define any reward and simply provides a way to remote control an AI agent via rpc (w/ a python wrapper for convenience). I decided not to include a reward for a couple reasons (**tl;dr** - basic reward function is trivial and I want to minimize required changes to 0 AD when exploring reward shaping, action spaces, and state spaces):
- the observation already includes win/lose information (player state is set to "active", "won", or "defeated"). As a result, the default reward can be computed with simply checking this field for the current player.
- Computing the reward in 0 AD itself may result in reward shaping efforts to be implemented in 0 AD as well. A similar case could be made for providing more restricted action/state spaces. This would result in experiments requiring custom builds of 0 AD and would require recompiling 0 AD (making reproducibility and exploring combinations of rewards/actions/state spaces more cumbersome). Ideally, if this PR can be integrated into 0 AD, then the installation would be the same for all experiments (with potentially the exception that they may require custom mods after the initial installation). As these changes should add relatively little amounts of overhead to implement in python, I opted for keeping a simple, clean interface in 0 AD that should be able to stay the same across experiments which may want to customize action spaces, state spaces, and reward shaping.

**Comment by stevenhutt** [comment]
How would the RL work? Typically it needs hundreds of thousands of games to learn. Do you envisage the RL playing itself. It might be interesting if players agreed to uploading their gameplay and somehow using this as the training set, but I admit I haven't thought this through carefully.

**Comment by brollb** [comment]
This PR is meant to simply provide the means to explore RL/ML agents (or simply prototype agents using classical techniques in Python). As you said, most RL algorithms would need to play many games before it is able to learn how to play effectively (especially in a game as complex as 0 AD!). This PR tries to provide the necessary features to 0 AD so this type of exploration is feasible.

In particular, this PR enables 0 AD to be controlled via an external program. This allows an external script to [start games](https://github.com/0ad/0ad/pull/25/files#diff-31ca83971d39c5e70a3c2f99acdec9cfR17-R18), [apply actions in lock-step](https://github.com/0ad/0ad/pull/25/files#diff-31ca83971d39c5e70a3c2f99acdec9cfR33), and [query any relevant info](https://github.com/0ad/0ad/pull/25/files#diff-47868c7176e54cf350b4b5400133c7dbR76-R101) about units. This makes it possible to train RL agents as they can play hundreds of thousands of games (either against themselves or against the built-in AI). There is a very simple example demonstrating the ability to control multiple armies [here](https://github.com/0ad/0ad/pull/25/files#diff-742d11b2b2c6145c273502bdaaa5cc17).

I definitely agree that it would be great to have players donate their own replays to leverage for an offline approach (which could help us warm-start an RL model). There is a [relevant post](https://wildfiregames.com/forum/index.php?/topic/25610-public-domain-dataset-for-machine-learning/) on the forum about aggregating human replays but it would be great to have an official place for something like this!

**Comment by stevenhutt** [comment]
It's good to know all the 'pieces' are there for RL, if not the training data, so that can do some basic experiments. And agreed the game is complex for full RL modelling, but maybe a component of decision making could be accessible. But it's hard for enthusiasts to get RL results like Google not because we aren't as smart but because we don't have access to 100s of GPUs. 

I suppose I would need to compile 0ad from source to access PR?

**Comment by brollb** [comment]
The quickest way would be to use a [docker image](https://hub.docker.com/r/brollb/0ad) I made though I have mostly just been using it myself so there isn't a lot of documentation. It starts the RPC interface listening on port 50050 and the user is called "user" so you would probably want to run it with:
```
docker run -d -p 50051:50050 -v /path/to/replays/dir:/home/user/.local/share/0ad/replays brollb/0ad
```
where `/path/to/replays/dir` is where you would like to store the replays locally (must be globally writeable). Then you should be able to run the sample (using python 3.6 or greater) with:
```
python source/tools/clients/python/samples/multi-army-control.py
```
After playing some games, you should be able to load the replays using vanilla 0 AD (as they will be stored in whatever path was used in place of `/path/to/replays/dir`). Don't forget to kill the docker container with `docker kill <container id>` where `<container id>` is the hash printed after starting the container originally.

If you don't want to use docker, then you will need to recompile 0 AD. You also need recent versions of protobuf (including the compiler, `protoc`) and grpc [installed](https://github.com/0ad/0ad/pull/25/files#diff-c32fe1930308672509be42849f03aa6aR724-R734). I do most of my development on arch and still need to make it easier on ubuntu and distros with older versions of protobuf in the repos (I am using protobuf v3.7.0-1 and grpc v1.19.1-1).

**Comment by iolalla** [comment]
I got the code running and looks promising, interesting and definetly fun.

Regarding the reward, simplest could be maximize number of kills ? minimize number of enemies ? and could be calculated in the multi-army-control.py, so I think this could work.

Definetly an interesting branch of the game! Thanks for sharing!

**Comment by brollb** [comment]
@iolalla - Happy you enjoyed it!

Yeah, the most obvious reward function would be simply 1, if the player wins the scenario (available in the game state JSON) and -1 if the player loses (0 otherwise). These rewards would be applicable then for any different game style (ie, victory conditions supported by 0 AD) but they would be sparse which might make it a bit harder to train an agent. 

Depending upon the scenario, the reward functions that you mentioned could certainly be helpful with reward shaping. However, I decided not to include a reward or OpenAI Gym-like action/state space as both of these may be things that people may like to explore and thought it would be better to simply expose the required fundamental capabilities in 0 AD and allow people to add custom logic on the python side (reward functions, action/observation spaces, etc). This way exploration into different reward functions, action spaces, and observation spaces all can use the same build of 0 AD.

An additional perk of this approach is that it also results in a pretty simple interface which enables people to develop AI using classical techniques easily. Since it is using GRPC, these agents could also be in any supported language such as C#, Java, NodeJS, etc.

**Comment by StanleySweet** [comment]
I guess now that it was merged you can close it :)
